google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/ConnectionImpl.java

```diff
@@ -0,0 +1,922 @@
+  }
+
+  /** The combination of all transaction modes and batch modes. */
+  enum UnitOfWorkType {
+    READ_ONLY_TRANSACTION {
+      @Override
+      TransactionMode getTransactionMode() {
+        return TransactionMode.READ_ONLY_TRANSACTION;
+      }
+    },
+    READ_WRITE_TRANSACTION {
+      @Override
+      TransactionMode getTransactionMode() {
+        return TransactionMode.READ_WRITE_TRANSACTION;
+      }
+    },
+    DML_BATCH {
+      @Override
+      TransactionMode getTransactionMode() {
+        return TransactionMode.READ_WRITE_TRANSACTION;
+      }
+    },
+    DDL_BATCH {
+      @Override
+      TransactionMode getTransactionMode() {
+        return null;
+      }
+    };
+
+    abstract TransactionMode getTransactionMode();
+
+    static UnitOfWorkType of(TransactionMode transactionMode) {
+      switch (transactionMode) {
+        case READ_ONLY_TRANSACTION:
+          return UnitOfWorkType.READ_ONLY_TRANSACTION;
+        case READ_WRITE_TRANSACTION:
+          return UnitOfWorkType.READ_WRITE_TRANSACTION;
+        default:
+          throw SpannerExceptionFactory.newSpannerException(ErrorCode.INVALID_ARGUMENT,
+              "Unknown transaction mode: " + transactionMode);
+      }
+    }
+  }
+
+  private StatementExecutor.StatementTimeout statementTimeout = new StatementExecutor.StatementTimeout();
+  private boolean closed = false;
+
+  private final Spanner spanner;
+  private DdlClient ddlClient;
+  private DatabaseClient dbClient;
+  private boolean autocommit;
+  private boolean readOnly;
+
+  private UnitOfWork currentUnitOfWork = null;
+  /**
+   * The {@link ConnectionImpl#inTransaction} field is only used in autocommit mode to indicate that
+   * the user has explicitly started a transaction.
+   */
+  private boolean inTransaction = false;
+  /**
+   * This field is used to indicate that a transaction begin has been indicated. This is done by
+   * calling beginTransaction or by setting a transaction property while not in autocommit mode.
+   */
+  private boolean transactionBeginMarked = false;
+
+  private BatchMode batchMode;
+  private UnitOfWorkType unitOfWorkType;
+  private final Stack<UnitOfWork> transactionStack = new Stack<>();
+  private boolean retryAbortsInternally;
+  private final List<TransactionRetryListener> transactionRetryListeners = new ArrayList<>();
+  private AutocommitDmlMode autocommitDmlMode = AutocommitDmlMode.TRANSACTIONAL;
+  private TimestampBound readOnlyStaleness = TimestampBound.strong();
+
+  /** Create a connection and register it in the SpannerPool. */
+  ConnectionImpl(ConnectionOptions options) {
+    Preconditions.checkNotNull(options);
+    this.statementExecutor = new StatementExecutor(options.getStatementExecutionInterceptors());
+    this.spannerPool = SpannerPool.INSTANCE;
+    this.options = options;
+    this.spanner = spannerPool.getSpanner(options, this);
+    this.ddlClient = DdlClient.newBuilder().setDatabaseAdminClient(spanner.getDatabaseAdminClient())
```

Comment:
I'd move this logic to a private method, `createDdlClient`

Reply:
Moved to a separate method, and moved the assignment to the end of the constructor to ensure that all fields have been initialized.